Unit 11 : Zero-Shot Learning draft (#200)
johko/computer-vision-course: chapters/en/Unit 11  - Zero Shot Computer Vision/2.mdx

- **Animal with Attributes (AwA)**

Dataset to benchmark transfer-learning algorithms, in particular attribute based classification [6]. It consists of 30475 images of 50 animal classes with six feature representations for each image.  

- **Caltech-UCSD-Birds-200-2011（CUB）**

Dataset for fine-grained visual categorization task. It contains 11788 images of 200 subcategories of birds. Each image has 1 subcategory label, 15 part locations, 312 binary attributes and 1 bounding box. Also, ten-sentence descriptions for each image. were collected through Mechanical Turk by Amazon, and the descriptions are carefully constructed to not contain any subcategory information.  

- **Sun database（SUN）**

Scene categorization benchmarks. It consists of 130519 images of 899 categories. 

- **Attribute Pascal and Yahoo dataset（aPY）**

A coarse-grained dataset composed of 15339 images from 3 broad categories (animals, objects and vehicles), further divided into a total of 32 subcategories. 

- **ILSVRC2012/ILSVRC2010（ImNet-2）**

The ImageNet Large Scale Visual Recognition Challenge (ILSVRC) evaluates algorithms for object detection and image classification at large scale. 

> ## Reference

Reviewer: small nit, can we keep 2 empty lines between each section, can we remove extra 1 line here?

Author: Fixed thanks!

# Zero-shot Learning

Following the introductory chapter, we will next explain zero-shot learning (ZSL) in more details. This chapter is designed to cover:

> - 1. The definitions for various types of ZSL and the differences between them.   

Reviewer 2: ```suggestion
- The definitions for various types of ZSL and the differences between them.   
```

> - 2. An in-depth example of ZSL which employs semantic embeddings [1]. 

Reviewer 2: ```suggestion
- An in-depth example of ZSL which employs semantic embeddings [1]. 
```

In summary, semantic embeddings are central to ZSL, enabling models to extend their classification capabilities. This approach allows for a more flexible and scalable way to classify the vast amount of real-world categories without needing labeled datasets.

## Comparison to CLIP 

The relationship between ZSL and CLIP (Contrastive Language–Image Pre-training) [5] stems from their shared goal of enabling models to recognize and classify images of categories that were not present in the training data. However, CLIP represents a significant advancement and a broader application of the principles underlying ZSL, leveraging a novel approach to learning and generalization.

The relationship between CLIP and ZSL can be described as:

> - 1. Both ZSL and CLIP aim to classify images into classes that were not seen during training. However, while traditional ZSL approaches might rely on predefined semantic embeddings or attributes to bridge the gap between seen and unseen classes, CLIP directly learns from natural language descriptions, allowing it to generalize to a much broader range of tasks without the need for task-specific embeddings.

Reviewer 2: ```suggestion
- Both ZSL and CLIP aim to classify images into classes that were not seen during training. However, while traditional ZSL approaches might rely on predefined semantic embeddings or attributes to bridge the gap between seen and unseen classes, CLIP directly learns from natural language descriptions, allowing it to generalize to a much broader range of tasks without the need for task-specific embeddings.
```

> - 2. CLIP is a prime example of multi-modal learning, where the model learns from both textual and visual data. This approach aligns with ZSL where auxiliary information is used to improve classification performance. CLIP takes the concept further by directly learning from raw text and images, enabling it to understand and represent the relationships between visual content and descriptive language.

Reviewer 2: ```suggestion
- CLIP is a prime example of multi-modal learning, where the model learns from both textual and visual data. This approach aligns with ZSL where auxiliary information is used to improve classification performance. CLIP takes the concept further by directly learning from raw text and images, enabling it to understand and represent the relationships between visual content and descriptive language.
```

# Zero-shot Learning

Following the introductory chapter, we will next explain zero-shot learning (ZSL) in more details. This chapter is designed to cover:
- 1. The definitions for various types of ZSL and the differences between them.   
- 2. An in-depth example of ZSL which employs semantic embeddings. 

> 

Reviewer 3: ```suggestion
```

## Zero-shot Learning vs. Generalized Zero-shot Learning

Zero-shot learning and generalized zero-shot learning (GZSL) belong to a machine learning algorithm type, where the image classification model needs to classify labels not included in training. ZSL and GZSL are very similar, and the main difference is how the model is evaluated [2].

For ZSL, the model is purely evaluated for its capability to classify images of unseen classes - only observations of unseen classes are included in ZSL testing dataset. For GZSL, the model is evaluated on both seen classes and unseen classes - which is considered closer to real-world use cases. Overall, GZSL is more challenging because the model needs to determine if an observation belongs to a novel class or a known class. 

> 

Reviewer 3: ```suggestion
```

## Inductive Zero-shot Learning vs. Transductive Zero-shot Learning

Based on the type of training data, there are two kinds of zero-shot learning:

In inductive ZSL, the model is trained exclusively on datasets containing only seen classes, without access to any data from the unseen classes. The learning process focuses on extracting and generalizing patterns from the training data, which are then applied to classify instances of unseen classes. The approach assumes a clear separation between seen and unseen data during training, emphasizing the model's ability to generalize from the training data to the unseen classes. 

Transductive ZSL differs by allowing the model to have access to some information about the unseen classes during training, typically the attributes or unlabeled examples of the unseen classes, but not the labels. This approach leverages additional information about the structure of the unseen data to train a more generalizable model. 

In next section, we will follow the main concept of a classic research paper from Google [1], and give an example for inductive ZSL. 

> 

Reviewer 3: ```suggestion
```

### Enable Zero-shot Learning with Semantic Embeddings  

During training, a ZSL model learns to associate the visual features of images from seen classes with their corresponding semantic embeddings. The objective is to minimize the distance between the projected visual features of an image and the semantic embedding of its class. This process helps the model to learn the correspondence between images and semantic information.

Since the model has learned to project image features onto semantic space, it can attempt to classify images of unseen classes by projecting their image features into the same space and comparing them to the embeddings of unseen classes. For an image of an unseen class, the model calculates its projected embedding and then searches for the nearest semantic embedding among the unseen classes. The unseen class with nearest embedding is the predicted label for the image.

In summary, semantic embeddings are central to ZSL, enabling models to extend their classification capabilities. This approach allows for a more flexible and scalable way to classify the vast amount of real-world categories without needing labeled datasets.

> 

Reviewer 3: ```suggestion
```

The relationship between CLIP and ZSL can be described as:

- 1. Both ZSL and CLIP aim to classify images into classes that were not seen during training. However, while traditional ZSL approaches might rely on predefined semantic embeddings or attributes to bridge the gap between seen and unseen classes, CLIP directly learns from natural language descriptions, allowing it to generalize to a much broader range of tasks without the need for task-specific embeddings.

- 2. CLIP is a prime example of multi-modal learning, where the model learns from both textual and visual data. This approach aligns with ZSL where auxiliary information is used to improve classification performance. CLIP takes the concept further by directly learning from raw text and images, enabling it to understand and represent the relationships between visual content and descriptive language.

> 

Reviewer 3: ```suggestion
```

- **Caltech-UCSD-Birds-200-2011（CUB）**

Dataset for fine-grained visual categorization task. It contains 11788 images of 200 subcategories of birds. Each image has 1 subcategory label, 15 part locations, 312 binary attributes and 1 bounding box. Also, ten-sentence descriptions for each image. were collected through Mechanical Turk by Amazon, and the descriptions are carefully constructed to not contain any subcategory information.  

- **Sun database（SUN）**

First large-scale scene attribute database. The dataset consists of 130519 images with 899 categories which can be used in high-level scene understanding and fine-grained scene recognition. 

- **Attribute Pascal and Yahoo dataset（aPY）**

A coarse-grained dataset composed of 15339 images from 3 broad categories (animals, objects and vehicles), further divided into a total of 32 subcategories. 

- **ILSVRC2012/ILSVRC2010（ImNet-2）**

The ImageNet Large Scale Visual Recognition Challenge (ILSVRC) evaluates algorithms for object detection and image classification at large scale [8]. 

> 

Reviewer 3: ```suggestion
```

# Zero-shot Learning

> 

Reviewer 2: ```suggestion
```

> ## 1. ZSL vs. Generalized ZSL

Reviewer 2: ```suggestion
## Zero-shot Learning vs. Generalized Zero-shot Learning
```

> Zero-shot learning (ZSL) and generalized zero-shot learning (GZSL) both belong to a type of machine learning algorithm, where the image classification model needs to classify labels not included in training. ZSL and GZSL are very similar, and the main difference is how the model is evaluated.

Reviewer 2: ```suggestion
Zero-shot learning (ZSL) and generalized zero-shot learning (GZSL) belong to a machine learning algorithm type, where the image classification model needs to classify labels not included in training. ZSL and GZSL are very similar, and the main difference is how the model is evaluated.
```

> For ZSL, the model is purely evaluated for its capability to classify images of unseen classes - only observations of unseen classes are included in ZSL testing dataset. As for GZSL, the model is evaluated on both seen classes and unseen classes - which is considered more practical and closer to real-world use cases. Overall, GZSL is more challenging, because the model needs to determine if an observation belongs to a novel class or a known class. 

Reviewer 2: ```suggestion
For ZSL, the model is purely evaluated for its capability to classify images of unseen classes - only observations of unseen classes are included in ZSL testing dataset. As for GZSL, the model is evaluated on both seen classes and unseen classes - which is considered more practical and closer to real-world use cases. Overall, GZSL is more challenging because the model needs to determine if an observation belongs to a novel class or a known class. 
```

As described in the previous section, developing a successful ZSL model requires more than just images and class labels. It is nearly impossible to classify unseen classes based on images alone. ZSL utilizes semantic embeddings to help classify images from unseen classes. 

### Semantic Embeddings 

> Semantic embeddings are vector representations of semantic information - the meaning and interpretation of textual data. For example, the information transferred with spoken text is a type of semantic information. Semantic information does not include only the direct meanings of words or sentences, but also the contextual and cultural implications.

Reviewer 2: ```suggestion
Semantic embeddings are vector representations of semantic information - the meaning and interpretation of textual data. For example, the information transferred with spoken text is a type of semantic information. Semantic information does not include only the direct meanings of words or sentences but also the contextual and cultural implications.
```

> Embeddings refer to the process of mapping semantic information into vectors of real numbers. Semantic embeddings are often learned with unsupervised machine learning models, such as Word2Vec [1] or GloVe [2]. All types of textual information such as words, phrases, or sentences can be transformed into numerical vectors based on set procedures. Semantic embeddings describe words in a high-dimensional space where the distance and direction between words reflect their semantic relationships. This enables machines to understand the usage, synonyms, and context of each word by mathematical operations on the word embeddings. 

Reviewer 2: ```suggestion
Embeddings refer to the process of mapping semantic information into vectors of real numbers. Semantic embeddings are often learned with unsupervised machine learning models, such as Word2Vec [1] or GloVe [2]. All types of textual information, such as words, phrases, or sentences can be transformed into numerical vectors based on set procedures. Semantic embeddings describe words in a high-dimensional space where the distance and direction between words reflect their semantic relationships. This enables machines to understand the usage, synonyms, and context of each word by mathematical operations on the word embeddings. 
```

I think you can directly give markdown links instead of giving citation like this

> 

Author: Thank you I added a section to explain how the embeddings are used in ZSL.

> ## 2. Types of ZSL Based on Training Data

Reviewer 2: ```suggestion
## Types of ZSL Based on Training Data
```

> Based on the types of training data, there are two kinds of zero-shot learning:

Reviewer 2: ```suggestion
Based on the types of training data, there are two kinds of zero-shot learning.
```

> ### Inductive zero shot

Reviewer 2: ```suggestion
### Inductive  Zero-shot Learning
```

> Training data includes labeled images of seen classes, and semantic descriptions/attributes of both seen and unseen classes. The main idea is to establish projections from observed image space to semantic latent space, and the model can identify unseen classes at inference time.

Reviewer 2: ```suggestion
Training data includes labeled images of seen classes and semantic descriptions/attributes of both seen and unseen classes. The main idea is to establish projections from observed image space to semantic latent space, and the model can identify unseen classes at inference time.
```

> ### Transductive zero shot

Reviewer 2: ```suggestion
### Transductive Zero-shot Learning
```

> Training data includes labelled images of seen classes and unlabelled images of unseen classes, and semantic descriptions or attributes of both seen and unseen classes. The usecase for transductive learning is when we do not have enough labeling resource to label all images.

Reviewer 2: ```suggestion
Training data includes labeled images of seen classes, and unlabeled images of unseen classes, and semantic descriptions or attributes of both seen and unseen classes. The use case for transductive learning is when we do not have enough labeling resources to label all images.
```

> ## 3. Zero Shot Learning Example 

Reviewer 2: ```suggestion
## Zero-shot Learning Example 
```

> DeViSE - A Deep Visual-Semantic Embedding Model

Reviewer 2: Would you like to add this as a separate page?

Author: Thank you for the suggestion! The ZSL example - DeViSE section is removed,  and we plan to add as a separate notebook.

First, an image classification model - the core visual model - is trained with labels (Fig 1). Next, a transformation layer after the output of core visual model is trained along with the embedding vectors of image labels to increase the dot product similarity. At inference time, when a new image arrives, the output of core visual model and the transformation layer is first computed, and the nearest label in semantic embedding space will be the predicted label for the image.  

- Results

On the flat metric, the softmax baseline shows higher accuracy for k = 1, 2. However,  for k = 5, 10 the 1000-D DeViSE model reached similar accuracy as baseline. The softmax model was expected to be the best model, given its training objective is aligned with the evaluation metric. It was surprising that the performance was so similar to the softmax performance.

|Model|dim|Flat Hit @ 1 (%)|Flat Hit @ 2 (%)|Flat Hit @ 5 (%)|Flat Hit @ 10 (%)|
|----------------|---|----------------|----------------|----------------|-----------------|
|Softmax baseline|N/A|55.6%|67.4%|78.5%|85.0%|
|DeViSE|500|53.2%|65.2%|76.7%|83.3%|
|DeViSE|1000|54.9%|66.9%|78.4%|85.0%|

## 4. Benchmarks and Evaluation

> The benchmarks and standard datasets are:

Author: Thank you for the suggestion! I added a short section to introduce the evaluation datasets.

> - 1. Animal with Attributes (AwA）

Reviewer 2: ```suggestion
- **Animal with Attributes (AwA)**
```
Can you apply the same suggestion to rest of the datasets?
